pkg/pillar/cmd/zedagent/handlenetworkinstance.go

```diff
@@ -696,3 +701,128 @@ func publishInfoToZedCloud(UUID string, infoMsg *zinfo.ZInfoMsg, iteration int)
+	flows := cast.CastFlowStatus(statusArg)
+	if flows.Key() != key {
+		log.Errorf("handleAppFlowMonitorModify key/UUID mismatch %s vs %s; ignored %+v\n",
+			key, flows.Key(), flows)
+		return
+	}
+
+	flowIteration++
+	// encoding the flows with protobuf format
+	pflows := protoEncodeAppFlowMonitorProto(flows)
+
+	// send protobuf to zedcloud
+	sendFlowProtobuf(pflows, flowIteration)
+}
+
+func handleAppFlowMonitorDelete(ctxArg interface{}, key string,
+	statusArg interface{}) {
+
+	log.Infof("handleAppFlowMonitorDelete(%s)\n", key)
+	flows := cast.CastFlowStatus(statusArg)
+	if flows.Key() != key {
+		log.Errorf("handleAppFlowMonitorDelete key/UUID mismatch %s vs %s; ignored %+v\n",
+			key, flows.Key(), flows)
+		return
+	}
+	log.Infof("handleAppFlowMonitorDelete(%s) done\n", key)
+}
+
+func protoEncodeAppFlowMonitorProto(ipflow types.IPFlow) *zflow.FlowMessage {
+
+	pflows := new(zflow.FlowMessage)
+	pflows.DevId = ipflow.DevID.String()
+
+	// ScopeInfo fill in
+	pScope := new(zflow.ScopeInfo)
+	pScope.Uuid = ipflow.Scope.UUID.String()
+	pScope.Intf = ipflow.Scope.Intf
+	pScope.LocalIntf = ipflow.Scope.Localintf
+	pScope.NetInstUUID = ipflow.Scope.NetUUID.String()
+	pflows.Scope = pScope
+
+	// get the ip flows from the input
+	for _, rec := range ipflow.Flows {
+		prec := new(zflow.FlowRecord)
+
+		// IpFlow fill in
+		pIpflow := new(zflow.IpFlow)
+		pIpflow.Src = rec.Flow.Src.String()
+		pIpflow.Dest = rec.Flow.Dst.String()
+		pIpflow.SrcPort = int32(rec.Flow.SrcPort)
+		pIpflow.DestPort = int32(rec.Flow.DstPort)
+		pIpflow.Protocol = int32(rec.Flow.Proto)
+		prec.Flow = pIpflow
+
+		prec.Inbound = rec.Inbound
+		prec.AclId = rec.ACLID
+		// prec.AclName =
+		pStart := new(timestamp.Timestamp)
+		pStart = timeNanoToProto(rec.StartTime)
+		prec.StartTime = pStart
+		pEnd := new(timestamp.Timestamp)
+		pEnd = timeNanoToProto(rec.StopTime)
+		prec.EndTime = pEnd
+		prec.TxBytes = rec.TxBytes
+		prec.TxPkts = rec.TxPkts
+		prec.RxBytes = rec.RxBytes
+		prec.RxPkts = rec.RxPkts
+		pflows.Flows = append(pflows.Flows, prec)
+	}
+
+	// get the ip DNS records from the input
+	for _, dns := range ipflow.DNSReqs {
+		pdns := new(zflow.DnsRequest)
+		pdns.HostName = dns.HostName
+		for _, address := range dns.Addrs {
+			pdns.Addrs = append(pdns.Addrs, address.String())
+		}
+		dnsTime := new(timestamp.Timestamp)
+		dnsTime = timeNanoToProto(dns.RequestTime)
+		pdns.RequestTime = dnsTime
+		pdns.AclNum = dns.ACLNum
+		pflows.DnsReqs = append(pflows.DnsReqs, pdns)
+	}
+
+	return pflows
+}
+
+func sendFlowProtobuf(pflows *zflow.FlowMessage, iteration int) {
+	data, err := proto.Marshal(pflows)
+	if err != nil {
+		log.Errorf("FlowStats: SendFlowProtobuf proto marshaling error %v", err) // XXX change to fatal
+	}
+
+	buf := bytes.NewBuffer(data)
+	size := int64(proto.Size(pflows))
+	flowlogURL := serverNameAndPort + "/" + flowlogAPI
+	const return400 = false
+	_, _, cf, err := zedcloud.SendOnAllIntf(zedcloudCtx, flowlogURL,
```

Comment:
Need to decide what to do if a send fails and document that. Will a subsequent send include all what was in the failed send, in which case you can ignore the failure. That is the case for the current metric messages. If not you need to do defer handling the way we do for info messages.

Reply:
I'll put a comment there now, and come back to add the 'defer resend' action code